documentation/smg/17_using_pixel_data.md

```diff
@@ -0,0 +1,394 @@
+the number of workers, the number of pixel, the page size.
+Each worker can receive its start and end page index from the head node,
+and operate on those pixels.
+
+## The PixelData Class
+
+### Principles
+
+- It operates with "pages" of cached data.
+  - The page size is specified on construction (in bytes).
+  - Only a page size worth of pixels is loaded at any one time.
+  - If all pixels fit in one page, all pixels are loaded.
+
+- Temporary files hold changes to pixel data.
+  - When pixel data is changed, the changes are written to a temporary file
+    when the current page is cleared.
+    This most often occurs when the next page is loaded.
+  - Changes are _never_ written to the SQW file used in construction.
+  - A logical array is held on the object that tracks which pages are "dirty".
+  - If a page is "clean" the data is loaded from the SQW file, if the page is
+    "dirty", it is loaded from a temporary file.
+  - Each page has its own temporary file.
+
+- It is a
+  [`handle` class](https://www.mathworks.com/help/matlab/matlab_oop/handle-objects.html).
+  - To copy a `PixelData` object, you must explicitly call the `copy` method.
+    This copy can be expensive, as all temporary files are also copied.
+  - There is no "lazy-copy-on-assignment" as exists for Matlab value classes.
+
+  The main reason for using a handle class was due to the file-backed nature of
+  the object.
+  As the data resides in a file, copies of `PixelData` objects are not copies of
+  the data, and thus, the objects are not independent.
+
+  Using a handle class is also convenient when working with setters.
+  Handle classes allow you to do the following:
+
+  ```matlab
+  >> sqw_obj.data.pix.signal = zeros(1, pix.num_pixels);
+  ```
+
+  Whereas, using a value class, one must return and assign the class instance
+  for the setter to take effect:
+
+  ```matlab
+  >> sqw_obj.data.pix = sqw_obj.data.pix.signal = zeros(1, pix.num_pixels);
+  ```
+
+- Data is loaded on access.
+  - No data is loaded from file on construction.
+  - All public data accessors _must_ check that data is loaded before returning data.
+
+  This is advantageous as it prevents performing potentially expensive
+  IO operations when pixels are not used.
+  For example, a user may wish to load an SQW object and plot the image data.
+  This does not require pixels, so time and memory is wasted loading what could
+  be gigabytes of data.
+  Note that this also means it is free to move to a given page of data after
+  constructing the object.
+
+- Pixel data are held in memory as `double`s.
+  This is in contrast to in file (including temporary page files),
+  where pixels are stored as `single`.
+  See [ADR 15](../adr/0015-store-pixel-data-in-single-precision.md).
+
+- `PixelData` objects do not know about their binning.
+  - The binning is defined by the `sqw` object's image.
+
+### Getters/Setters
+
+All getters return data in-memory.
+All setters must set data that is in memory.
```

Comment:
Is it an error to fail to do so?

Reply:
The values you set must be numeric, so it would be an error to pass any of the setters anything other than an in-memory array.

Although, no `set_pixels` method exists at the moment (`set_data` was only recently added), but that could in theory set pixel data from a file-backed `PixelData` object.

Should I remove the setters part of this in case a `set_pixels` method is ever required?

Reply:
I think the getters bit is clarified below but useful. I just think it would be good to note what happens if you attempt to set pixels out of current page, or set pixels out of range? If it's an error, note "Anything else will result in an error"